### PR TITLE
support required fields without ... Fixes #34

### LINKED ISSUE #34: typerror in example: Incompatible types in assignment (expression has type "ellipsis", variable has type "int")
with:

```python
from datetime import datetime
from pydantic import BaseModel

class UserModel(BaseModel):
    id: int = ...
    name = 'John Doe'
    signup_ts: datetime = None

external_data = {'id': '123', 'signup_ts': '2017-06-01 12:22'}
user = UserModel(**external_data)
print(user)
# > UserModel id=123 name='John Doe' signup_ts=datetime.datetime(2017, 6, 1, 12, 22)
print(user.id)
# > 123
```
I get:
```
error: Incompatible types in assignment (expression has type "ellipsis", variable has type "int")
error: Incompatible types in assignment (expression has type None, variable has type "datetime")
```

### PR DESCRIPTION
**Pull request by graingert**

### CODE REVIEW

**Comment by samuelcolvin** [comment]
Why is this required? In what context does ellipsis not work?

**Comment by graingert** [comment]
For the example to typecheck under mypy for me I needed to create this patch, and correctly set the optional fields:

```python
from datetime import datetime
from pydantic import BaseModel, REQUIRED

class UserModel(BaseModel):
    id: int = REQUIRED
    name = 'John Doe'
    signup_ts: Optional[datetime] = None
```

Because None is not an instance of datetime, and Ellipsis is not an instance of int, mypy will never validate: `id: int = ...` or `signup_ts: Optional[datetime] = None`

**Comment by graingert** [comment]
@samuelcolvin I really don't know how `id: int = ...` could ever work...?

**Comment by graingert** [comment]
Also, personally, I find `id: int = REQUIRED` much more readable.

**Comment by samuelcolvin** [comment]
It works absolutely fine with python, just doesn't play nicely with mypy. mypy is a cool project, but it's far from common use and in my experience doesn't provide that much value in development.

I'd be happy to accept a change like this if it makes it clearer and allows pydantic to play nicely with mypy. But we would need some tests to confirm that code using this model actually passed with mypy.

**Comment by graingert** [comment]
@samuelcolvin cool, I'll make a PR to run the typechecker as part of the tests.

File: docs/index.rst [hunk 1]
```diff
@@ -24,8 +24,9 @@ Simple example:
 
 What's going on here:
 
-* ``id`` is of type int, the elipisis tells pydantic this field is required. Strings, bytes or floats will be
+* ``id`` is of type int, the ellipisis tells pydantic this field is required. Strings, bytes or floats will be
```
Comment (minus7) [inline docs/index.rst]: There's still a typo in ellipsis here and two lines below

Comment (samuelcolvin) [inline docs/index.rst]: i think this is fixed on master anyway.

**Comment by minus7** [comment]
Is there a specific reason `...` or `REQUIRED` is needed at all? You can have annotations without assigning values, e.g. `id: int` works just fine and is available through `cls.__annotations__`, you just can't read the attribute through `cls.id`/`obj.id` as long as it has not been set.

**Comment by samuelcolvin** [comment]
@minus7 I think you make a great point, I only noticed this syntax yesterday.

If it doesn't have a performance impact I'd be very open to removing the ellipsis thing altogether.

**Comment by samuelcolvin** [comment]
especially if it works with mypy without having to switch to `REQUIRED`

**Comment by samuelcolvin** [comment]
Thanks a lot, I've re-implemented this in #41 based on comments from #34 but this was a useful start.

**Pull request closed without merging**